File: hearingclinic/public/js/Sales_Invoice/sales_invoice_partial_payment.js
/**
 * Sales Invoice Partial Payment Client Script
 * Version: 2024-11-29 Final
 * File: hearingclinic/public/js/sales_invoice_partial_payment.js
 */

frappe.ui.form.on('Sales Invoice', {
    refresh: function(frm) {
        // Add "Record Final Payment" button for partly paid POS invoices
        if (frm.doc.docstatus === 1 && 
            frm.doc.outstanding_amount > 0 && 
            frm.doc.is_pos === 1) {
            
            frm.add_custom_button(__('Record Final Payment'), function() {
                record_final_payment(frm);
            }, __('Actions'));
        }
        
        // Display payment history for submitted invoices
        if (frm.doc.docstatus === 1) {
            display_payment_history(frm);
        }
    },
    
    is_pos: function(frm) {
        // Reset partial payment amount when POS checkbox changes
        if (!frm.doc.is_pos) {
            frm.set_value('custom_partial_payment_amount', 0);
        }
    },
    
    custom_partial_payment_amount: function(frm) {
        // Validate partial payment amount
        if (frm.doc.custom_partial_payment_amount) {
            if (frm.doc.custom_partial_payment_amount > frm.doc.grand_total) {
                frappe.msgprint(__('Partial payment amount cannot exceed invoice total'));
                frm.set_value('custom_partial_payment_amount', 0);
                return;
            }
            if (frm.doc.custom_partial_payment_amount < 0) {
                frappe.msgprint(__('Partial payment amount must be positive'));
                frm.set_value('custom_partial_payment_amount', 0);
                return;
            }
        }
    },
    
    before_save: function(frm) {
        // Store the payment amount for this transaction before submission
        if (frm.doc.is_pos && frm.doc.docstatus === 0) {
            if (frm.doc.custom_partial_payment_amount && frm.doc.custom_partial_payment_amount > 0) {
                frm.set_value('custom_amount_paid_this_transaction', frm.doc.custom_partial_payment_amount);
            } else {
                // Full payment - store the grand total
                frm.set_value('custom_amount_paid_this_transaction', frm.doc.grand_total);
            }
        }
    }
});

function record_final_payment(frm) {
    let d = new frappe.ui.Dialog({
        title: __('Record Final Payment'),
        fields: [
            {
                fieldname: 'outstanding_amount',
                label: __('Outstanding Amount'),
                fieldtype: 'Currency',
                default: frm.doc.outstanding_amount,
                read_only: 1,
                options: frm.doc.currency
            },
            {
                fieldname: 'payment_amount',
                label: __('Payment Amount'),
                fieldtype: 'Currency',
                default: frm.doc.outstanding_amount,
                reqd: 1,
                options: frm.doc.currency
            },
            {
                fieldname: 'mode_of_payment',
                label: __('Mode of Payment'),
                fieldtype: 'Link',
                options: 'Mode of Payment',
                reqd: 1,
                get_query: function() {
                    return {
                        filters: {
                            'enabled': 1
                        }
                    };
                }
            },
            {
                fieldname: 'posting_date',
                label: __('Posting Date'),
                fieldtype: 'Date',
                default: frappe.datetime.get_today(),
                reqd: 1
            },
            {
                fieldname: 'reference_no',
                label: __('Reference No'),
                fieldtype: 'Data',
                description: __('Cheque/Reference Number')
            },
            {
                fieldname: 'reference_date',
                label: __('Reference Date'),
                fieldtype: 'Date',
                depends_on: 'eval:doc.reference_no'
            },
            {
                fieldname: 'remarks',
                label: __('Remarks'),
                fieldtype: 'Small Text',
                default: __('Final payment for invoice {0}', [frm.doc.name])
            }
        ],
        primary_action_label: __('Record Payment'),
        primary_action(values) {
            // Validate payment amount
            if (values.payment_amount > frm.doc.outstanding_amount) {
                frappe.msgprint(__('Payment amount cannot exceed outstanding amount'));
                return;
            }
            if (values.payment_amount <= 0) {
                frappe.msgprint(__('Payment amount must be greater than zero'));
                return;
            }
            
            frappe.call({
                method: 'hearingclinic.hearingclinic.doc_events.sales_invoice_partial_payment.record_final_payment',
                args: {
                    invoice_name: frm.doc.name,
                    payment_amount: values.payment_amount,
                    mode_of_payment: values.mode_of_payment,
                    posting_date: values.posting_date,
                    reference_no: values.reference_no,
                    reference_date: values.reference_date,
                    remarks: values.remarks
                },
                freeze: true,
                freeze_message: __('Recording payment...'),
                callback: function(r) {
                    if (r.message) {
                        frappe.msgprint({
                            title: __('Payment Recorded'),
                            message: __('Payment Entry {0} created successfully', 
                                ['<a href="/app/payment-entry/' + r.message + '">' + r.message + '</a>']),
                            indicator: 'green'
                        });
                        frm.reload_doc();
                        d.hide();
                    }
                }
            });
        }
    });
    
    d.show();
}

function display_payment_history(frm) {
    // Remove existing payment history if present
    if (frm.fields_dict.custom_payment_history_html) {
        frm.fields_dict.custom_payment_history_html.$wrapper.empty();
    }
    
    frappe.call({
        method: 'hearingclinic.hearingclinic.doc_events.sales_invoice_partial_payment.get_payment_history',
        args: {
            invoice_name: frm.doc.name
        },
        callback: function(r) {
            if (r.message && r.message.payments !== undefined) {
                render_payment_history_table(frm, r.message);
            }
        }
    });
}

function render_payment_history_table(frm, data) {
    let html = `
        <div class="payment-history-section" style="margin-top: 15px; margin-bottom: 15px;">
            <h4 style="margin-bottom: 10px; color: #36414C;">Complete Payment History</h4>
            <table class="table table-bordered table-hover" style="margin-bottom: 10px;">
                <thead style="background-color: #F7FAFC;">
                    <tr>
                        <th style="width: 15%;">Date</th>
                        <th style="width: 25%;">Reference</th>
                        <th style="width: 20%;">Payment Method</th>
                        <th style="width: 15%; text-align: right;">Amount</th>
                        <th style="width: 25%;">Type</th>
                    </tr>
                </thead>
                <tbody>
    `;
    
    if (data.payments.length === 0) {
        html += `
            <tr>
                <td colspan="5" style="text-align: center; color: #8D99A6; padding: 20px;">
                    No payments recorded yet
                </td>
            </tr>
        `;
    } else {
        data.payments.forEach(function(payment) {
            let ref_link = payment.reference;
            if (payment.type === "Payment Entry") {
                ref_link = `<a href="/app/payment-entry/${payment.reference}" target="_blank">${payment.reference}</a>`;
            } else {
                ref_link = `<a href="/app/sales-invoice/${payment.reference}" target="_blank">${payment.reference}</a>`;
            }

            // Determine badge color based on payment type
            let badge_class = 'primary'; // Default for POS Payment
            if (payment.type === 'Payment Entry') {
                badge_class = 'success';
            } else if (payment.type === 'Value Add Card') {
                badge_class = 'warning'; // Yellow/orange badge for VAC
            }

            html += `
                <tr>
                    <td>${payment.date}</td>
                    <td>${ref_link}</td>
                    <td>${payment.mode_of_payment}</td>
                    <td style="text-align: right;">${format_currency(payment.amount, data.currency)}</td>
                    <td><span class="badge badge-${badge_class}">${payment.type}</span></td>
                </tr>
            `;
        });
    }
    
    html += `
                </tbody>
                <tfoot style="background-color: #F7FAFC; font-weight: bold;">
                    <tr>
                        <td colspan="3" style="text-align: right;">Total Paid:</td>
                        <td style="text-align: right;">${format_currency(data.total_paid, data.currency)}</td>
                        <td></td>
                    </tr>
    `;
    
    if (frm.doc.outstanding_amount > 0) {
        html += `
                    <tr>
                        <td colspan="3" style="text-align: right; color: #F56565;">Outstanding Balance:</td>
                        <td style="text-align: right; color: #F56565;">${format_currency(frm.doc.outstanding_amount, data.currency)}</td>
                        <td></td>
                    </tr>
        `;
    } else {
        html += `
                    <tr style="background-color: #C6F6D5;">
                        <td colspan="5" style="text-align: center; color: #22543D; font-weight: bold; padding: 10px;">
                            ✓ FULLY PAID
                        </td>
                    </tr>
        `;
    }
    
    html += `
                </tfoot>
            </table>
        </div>
    `;
    
    // Insert the HTML into the custom field
    if (frm.fields_dict.custom_payment_history_html) {
        frm.fields_dict.custom_payment_history_html.$wrapper.html(html);
    }
}

File: hearingclinic/hearingclinic/doc_events/sales_invoice_partial_payment.py
"""
Sales Invoice Partial Payment and Payment History API
Version: 2024-11-29 Final
File: hearingclinic/api/sales_invoice.py
"""

import frappe
from frappe import _
from frappe.utils import flt, nowdate, get_link_to_form, formatdate
from erpnext.accounts.doctype.payment_entry.payment_entry import get_payment_entry


@frappe.whitelist()
def record_final_payment(invoice_name, payment_amount, mode_of_payment, 
                        posting_date=None, reference_no=None, reference_date=None, remarks=None):
    """
    Record final payment for a Sales Invoice with partial payment
    
    Args:
        invoice_name: Name of the Sales Invoice
        payment_amount: Amount being paid
        mode_of_payment: Mode of Payment to use
        posting_date: Date of payment (defaults to today)
        reference_no: Cheque/reference number (optional)
        reference_date: Reference date (optional)
        remarks: Payment remarks (optional)
    
    Returns:
        str: Name of created Payment Entry
    """
    # Validate invoice exists and is submitted
    if not frappe.db.exists("Sales Invoice", invoice_name):
        frappe.throw(_("Sales Invoice {0} not found").format(invoice_name))
    
    invoice = frappe.get_doc("Sales Invoice", invoice_name)
    
    if invoice.docstatus != 1:
        frappe.throw(_("Sales Invoice must be submitted"))
    
    # Validate payment amount
    payment_amount = flt(payment_amount)
    outstanding_amount = flt(invoice.outstanding_amount)
    
    frappe.logger().debug(f"Recording payment: {payment_amount} for invoice {invoice_name}, outstanding: {outstanding_amount}")
    
    if payment_amount <= 0:
        frappe.throw(_("Payment amount must be greater than zero"))
    
    if payment_amount > outstanding_amount:
        frappe.throw(_("Payment amount cannot exceed outstanding amount of {0}").format(outstanding_amount))
    
    # Create Payment Entry using ERPNext's standard function
    pe = get_payment_entry("Sales Invoice", invoice_name, bank_amount=payment_amount)
    
    # Update payment entry with provided details
    pe.mode_of_payment = mode_of_payment
    pe.posting_date = posting_date or nowdate()
    pe.paid_amount = payment_amount
    pe.received_amount = payment_amount
    
    if reference_no:
        pe.reference_no = reference_no
    if reference_date:
        pe.reference_date = reference_date
    if remarks:
        pe.remarks = remarks
    
    # Update the references table with correct allocated amount
    for ref in pe.references:
        if ref.reference_doctype == "Sales Invoice" and ref.reference_name == invoice_name:
            ref.allocated_amount = payment_amount
            frappe.logger().debug(f"Set allocated amount to {payment_amount} for reference {ref.idx}")
    
    # Save and submit the payment entry
    pe.insert(ignore_permissions=True)
    frappe.logger().debug(f"Payment Entry {pe.name} inserted")
    
    pe.submit()
    frappe.logger().debug(f"Payment Entry {pe.name} submitted")
    
    # Update the Sales Invoice's custom field for printing
    frappe.db.set_value("Sales Invoice", invoice_name, 
                       "custom_amount_paid_this_transaction", payment_amount,
                       update_modified=False)
    
    # Commit the transaction to ensure all changes are saved
    frappe.db.commit()
    
    # Reload the invoice to get updated outstanding amount
    invoice.reload()
    
    frappe.logger().debug(f"After payment - Invoice outstanding: {invoice.outstanding_amount}, status: {invoice.status}")
    
    # Add a comment to the invoice
    invoice.add_comment("Comment", 
        _("Final payment of {0} recorded via {1}").format(
            frappe.format_value(payment_amount, {'fieldtype': 'Currency'}),
            get_link_to_form("Payment Entry", pe.name)
        )
    )
    
    return pe.name


@frappe.whitelist()
def get_payment_history(invoice_name):
    """
    Get complete payment history for a Sales Invoice
    Returns both POS payments and Payment Entries
    
    Args:
        invoice_name: Name of the Sales Invoice
    
    Returns:
        dict: Payment history with total paid
    """
    if not frappe.db.exists("Sales Invoice", invoice_name):
        return {"payments": [], "total_paid": 0}
    
    invoice = frappe.get_doc("Sales Invoice", invoice_name)
    payments = []
    
    # 1. Get POS payments (from the invoice itself)
    if invoice.payments:
        for payment in invoice.payments:
            # Determine payment type - special handling for Value Add Card
            payment_type = "Value Add Card" if payment.mode_of_payment == "Value Add Card" else "POS Payment"

            payments.append({
                "date": formatdate(invoice.posting_date),
                "reference": invoice.name,
                "mode_of_payment": payment.mode_of_payment,
                "amount": flt(payment.amount),
                "type": payment_type
            })
    
    # 2. Get Payment Entries linked to this invoice
    payment_entries = frappe.get_all(
        "Payment Entry Reference",
        filters={
            "reference_doctype": "Sales Invoice",
            "reference_name": invoice_name,
            "docstatus": 1
        },
        fields=["parent", "allocated_amount"]
    )
    
    for pe_ref in payment_entries:
        pe = frappe.get_doc("Payment Entry", pe_ref.parent)
        payments.append({
            "date": formatdate(pe.posting_date),
            "reference": pe.name,
            "mode_of_payment": pe.mode_of_payment,
            "amount": flt(pe_ref.allocated_amount),
            "type": "Payment Entry"
        })
    
    # Sort by date
    payments.sort(key=lambda x: x["date"])
    
    # Calculate total paid
    total_paid = sum([p["amount"] for p in payments])
    
    return {
        "payments": payments,
        "total_paid": total_paid,
        "currency": invoice.currency
    }


def override_pos_payment_amount(doc, method=None):
    """
    Hook function to override POS payment amount when partial payment is specified
    Called via validate hook

    Args:
        doc: Sales Invoice document
        method: Hook method name (unused)
    """
    if not doc.is_pos:
        return

    # Only process for unsaved/unsubmitted documents
    if doc.docstatus != 0:
        return

    # Skip for return invoices - ERPNext handles payment copying correctly
    if doc.is_return:
        return

    # Check if partial payment amount is specified
    partial_amount = flt(doc.get("custom_partial_payment_amount"))
    grand_total = flt(doc.grand_total)
    
    frappe.logger().debug(f"POS Payment Override - Partial: {partial_amount}, Grand Total: {grand_total}")
    
    if partial_amount > 0 and partial_amount < grand_total:
        # Validate partial amount
        if partial_amount > grand_total:
            frappe.throw(_("Partial payment amount cannot exceed invoice total"))
        
        # Override the payment amount in POS payments
        total_payment = 0
        for payment in doc.payments:
            # Set the payment amount to the partial amount
            payment.amount = partial_amount
            payment.base_amount = partial_amount
            total_payment += partial_amount
            frappe.logger().debug(f"Set payment {payment.idx} to {partial_amount}")
        
        # Update paid amount and change amount
        doc.paid_amount = partial_amount
        doc.base_paid_amount = partial_amount
        
        # Calculate change amount (should be 0 for partial payments normally)
        doc.change_amount = 0
        doc.base_change_amount = 0
        
        # Explicitly set outstanding amount
        doc.outstanding_amount = grand_total - partial_amount
        
        # Set the is_partial_payment flag
        doc.is_partial_payment = 1
        
        # Set status to indicate partial payment
        # ERPNext will calculate this in set_status() but we can hint it
        if hasattr(doc, 'status'):
            doc.status = 'Partly Paid'
        
        # Store this amount for printing
        doc.custom_amount_paid_this_transaction = partial_amount
        
        frappe.logger().debug(f"Final paid_amount: {doc.paid_amount}, outstanding: {doc.outstanding_amount}, is_partial_payment: 1, status: {doc.status}")
    else:
        # Full payment - ensure is_partial_payment is cleared
        doc.is_partial_payment = 0
        doc.custom_amount_paid_this_transaction = grand_total
        frappe.logger().debug(f"Full payment: {grand_total}, is_partial_payment: 0")


def ensure_partial_payment_flag(doc, method=None):
    """
    Hook to ensure is_partial_payment flag is correctly set after submission
    Called via on_submit hook
    """
    if not doc.is_pos:
        return

    # Skip for return invoices
    if doc.is_return:
        return

    # Check if there's an outstanding amount
    if flt(doc.outstanding_amount) > 0 and flt(doc.paid_amount) > 0:
        # There's both payment and outstanding - this is partial payment
        if not doc.is_partial_payment:
            frappe.logger().debug(f"Setting is_partial_payment flag on {doc.name}")
            frappe.db.set_value("Sales Invoice", doc.name, "is_partial_payment", 1, update_modified=False)
    else:
        # Full payment or no payment
        if doc.is_partial_payment:
            frappe.logger().debug(f"Clearing is_partial_payment flag on {doc.name}")
            frappe.db.set_value("Sales Invoice", doc.name, "is_partial_payment", 0, update_modified=False)